docs/architecture-decisions/2023-01-11_beaker_productionisation_review.md

```diff
@@ -0,0 +1,409 @@
+
+#### Before & After - user's perspective
+
+For most use cases, this should be a relatively small and probably imperceptible change.
+
+We believe there are two common usage scenarios that use the output of PyTEAL compilation currently:
+
+1. Output the `Application` via `Application.dump(...)`
+2. Interact with the `Application` by passing it to `ApplicationClient(app=..., ...)`.
+
+We propose maintaining those two scenarios without any immediate external changes, but internally:
+
+1. `Application.dump(...)` will call `Application.compile().dump()`, and potentially trigger a `DeprecationWarning` if we decide that we want users to always explicitly call compile.
+2. `ApplicationClient(app=..., ...)` will call `Application.compile()` and not retain any reference to `app`.
+
+To make use of scenarios 1 _and_ 2, or to control compilation parameters, a user should also be able to (for instance):
+
+```python
+app = Application(...)
+compiled_app: CompiledApplication = app.compile(...)
+compiled_app.dump(...)
+client = ApplicationClient(app=compiled_app, ...)
+```
+
+We suggest also potentially renaming `CompiledApplication.dump()`, perhaps to something along the lines of `serialize()`.
+
+The `compile()` call is actually a transpilation call (Beaker / PyTEAL transpilation -> TEAL), although it's called compile in PyTEAL so consideration should be made to either keep consistency with PyTEAL or use the more accurate `transpile()` (which also reduces confusion around the fact that you then have to call `compile` on algod to compile the TEAL to byte code before deployment).
+
+The exact details of what `CompiledApplication` will look like are TBD, but should be driven by the principles outlined in the "Why?" section above. Broadly, it stands to reason it would contain the approval and clear TEAL, the ABI spec and the app spec though at least.
+
+### (3) Renamings
+
+Renaming `version` parameter in `Application.__init__(version: int = pyteal.MAX_VERSION)` to (e.g.) `teal_version`, to be more explicit. Otherwise developers may be confused that it's the version of the specific smart contract. It may be desirable to allow `version` to continue to be specified for some time, but to raise a `DeprecationWarning`.
```

Comment:
this one I'm not so sure about but I get the general idea.

Reply:
Cool, do you mind elaborating on your thoughts on this one?

Reply:
I assume this is about how it should be `avm_version`, not `teal_version`, which we agree with and is being implemented.